### DISCUSSION TITLE
nixos-with-flakes/nixos-flake-configuration-explained

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#134] [Announcements]
# nixos-with-flakes/nixos-flake-configuration-explained

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/nixos-with-flakes/nixos-flake-configuration-explained

**Comment by AntonOfTheWoods** [comment]
Is `sudo nixos-rebuild switch --flake github:owner/repo#your-hostname` supposed to work with private repos? I just get 

```
error: unable to download 'https://api.github.com/repos/owner/repo/commits/HEAD': HTTP error 404
       response body:
       {
         "message": "Not Found",
         "documentation_url": "https://docs.github.com/rest/commits/commits#get-a-commit",
         "status": "404"
       }
```

I also tried `sudo nixos-rebuild switch --flake git+ssh://git@github.com/owner/repo#your-hostname` but here I just get

```
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
...
```
`git push` from the directory I'm in works fine.

**Reply by ryan4yin** [reply]
1. You'll need to use an SSH agent for authentication. Try running `ssh-add /path/to/your/key` to add your key first, and then give the switch another go.

2. Instead of using `sudo` straight up, try this command: `nixos-rebuild switch --flake git+ssh://git@github.com/owner/repo#your-hostname --use-remote-sudo`. It should do the trick.

Just to clarify the use of `--use-remote-sudo`, this flag is specifically designed to only attempt to gain `sudo` privileges during the `switch` phase of `nixos-rebuild`. By doing this, it allows the `build` phase to proceed with the permissions of the current user, which is important for accessing the `ssh-agent`.

**Comment by astrojhgu** [comment]
Does  `inputs.nixpkgs.url` equals the `nix-channel`?

**Reply by ryan4yin** [reply]
Yes.

**Comment by jjramsey** [comment]
Shouldn't `github:owner/repo#your-hostname` be surrounded by quotes? I'm surprised that `#` isn't interpreted as a comment character.

**Reply by whoamitty** [reply]
This is because there isn't space between both

```bash
$ echo hello world#yes I am here
hello world#yes I am here

$ echo hello world #yes I am here
hello world
```

**Comment by VitorCBSB** [comment]
This may be a super dumb question, but please bear with me, this is all completely new to me. I felt a bit confused about 2 things in the output parameter listing: "self" and "...". You explained self, but I don't see anything about the ellipsis. What does it mean and is it always needed or which situations does it apply?

**Reply by adrian-gierakowski** [reply]
... is to allow the caller to call the the function with an an arbitrary attribute set (which can have more attributes the the ones explicitly listed). Without it, you'll get an evaluation error or you call the function with attributes it doesn't expect.
